runtime/visualize.cc

```diff
@@ -0,0 +1,559 @@
+    *this << indent() << "consume(" << n->src << ");\n";
+    *this << indent() << "produce(" << n->dst << ");\n";
+  }
+
+  void visit(const allocate* n) override {
+    *this << indent() << "{ let " << n->sym << " = allocate('" << n->sym << "', " << static_cast<index_t>(n->elem_size)
+          << ", [\n";
+    *this << indent(2);
+    print_vector(n->dims, ",\n" + indent(2));
+    *this << "\n";
+    *this << indent(1) << "]);\n";
+    *this << n->body;
+    *this << indent(1) << "free(" << n->sym << ");\n";
+    *this << indent() << "}\n";
+  }
+
+  void visit(const make_buffer* n) override {
+    *this << indent() << "{ let " << n->sym << " = make_buffer(" << n->base << ", " << n->elem_size << ", [";
+    if (!n->dims.empty()) {
+      *this << "\n";
+      *this << indent(2);
+      print_vector(n->dims, ",\n" + indent(2));
+      *this << "\n";
+      *this << indent(1);
+    }
+    *this << "]);\n";
+    *this << n->body;
+    *this << indent() << "}\n";
+  }
+
+  void visit(const clone_buffer* n) override {
+    *this << indent() << "{ let " << n->sym << " = clone_buffer(" << n->src << ");\n";
+    *this << n->body;
+    *this << indent() << "}\n";
+  }
+
+  void visit(const crop_buffer* n) override {
+    *this << indent() << "{ let __" << n->sym << " = crop_buffer(" << n->sym << ", [";
+    if (!n->bounds.empty()) {
+      *this << "\n";
+      *this << indent(2);
+      print_vector(n->bounds, ",\n" + indent(2));
+      *this << "\n";
+      *this << indent();
+    }
+    *this << "]);\n";
+    *this << n->body;
+    *this << indent(1) << n->sym << " = __" << n->sym << ";\n";
+    *this << indent() << "}\n";
+  }
+
+  void visit(const crop_dim* n) override {
+    *this << indent() << "{ let __" << n->sym << " = crop_dim(" << n->sym << ", " << n->dim << ", " << n->bounds
+          << ");\n";
+    *this << n->body;
+    *this << indent(1) << n->sym << " = __" << n->sym << ";\n";
+    *this << indent() << "}\n";
+  }
+
+  void visit(const slice_buffer* n) override {
+    *this << indent() << "{ let __" << n->sym << " = slice_buffer(" << n->sym << ", {" << n->at << "});\n";
+    *this << n->body;
+    *this << indent(1) << n->sym << " = __" << n->sym << ";\n";
+    *this << indent() << "}\n";
+  }
+
+  void visit(const slice_dim* n) override {
+    *this << indent() << "{ let __" << n->sym << " = slice_dim(" << n->sym << ", " << n->dim << ", " << n->at << ");\n";
+    *this << n->body;
+    *this << indent(1) << n->sym << " = __" << n->sym << ";\n";
+    *this << indent() << "}\n";
+  }
+
+  void visit(const truncate_rank* n) override {
+    *this << indent() << "{ let __" << n->sym << " = truncate_rank(" << n->sym << ", " << n->rank << ");\n";
+    *this << n->body;
+    *this << indent(1) << n->sym << " = __" << n->sym << ";\n";
+    *this << indent() << "}\n";
+  }
+
+  void visit(const check* n) override { *this << indent() << "check(" << n->condition << ");\n"; }
+};
+
+const char* header =
```

Comment:
nit: would probably be easier to read/maintain if you used a C++ multiline string construct